express_relay/sdk/python/express_relay_utils/express_relay_client.py

```diff
@@ -0,0 +1,404 @@
+    ) -> UUID:
+        """
+        Submits a bid on an opportunity to the server via websocket.
+
+        Args:
+            opportunity_bid: An object representing the bid to submit on an opportunity.
+            subscribe_to_updates: A boolean indicating whether to subscribe to the bid status updates.
+            kwargs: Keyword arguments to pass to the HTTP post request.
+        Returns:
+            The ID of the submitted bid.
+        """
+        opportunity_bid_dict = opportunity_bid.model_dump()
+        if subscribe_to_updates:
+            result = await self.send_ws_message(
+                "post_opportunity_bid",
+                {
+                    "opportunity_id": opportunity_bid.opportunity_id,
+                    "amount": opportunity_bid.amount,
+                    "executor": opportunity_bid.executor,
+                    "permission_key": opportunity_bid.permission_key,
+                    "signature": opportunity_bid.signature,
+                    "valid_until": opportunity_bid.valid_until,
+                },
+            )
+            bid_id = UUID(result.get("id"))
+        else:
+            async with httpx.AsyncClient() as client:
+                resp = await client.post(
+                    urllib.parse.urlparse(self.server_url)
+                    ._replace(
+                        path=f"/v1/opportunities/{opportunity_bid.opportunity_id}/bids"
+                    )
+                    .geturl(),
+                    json=opportunity_bid_dict,
+                    **kwargs,
+                )
+
+            resp.raise_for_status()
+            bid_id = UUID(resp.json().get("id"))
+
+        return bid_id
+
+    async def ws_handler(
+        self,
+        opportunity_callback: (
+            Callable[[Opportunity], Coroutine[Any, Any, Any]] | None
+        ) = None,
+        bid_status_callback: (
+            Callable[[BidStatusUpdate], Coroutine[Any, Any, Any]] | None
+        ) = None,
+    ):
+        """
+        Continually handles new ws messages as they are received from the server via websocket.
+
+        Args:
+            opportunity_callback: An async function that serves as the callback on a new opportunity. Should take in one external argument of type Opportunity.
+            bid_status_callback: An async function that serves as the callback on a new bid status update. Should take in one external argument of type BidStatusUpdate.
+        """
+        if not self.ws:
+            raise ExpressRelayClientException("Websocket not connected")
+
+        async for msg in self.ws:
+            msg_json = json.loads(msg)
+
+            if msg_json.get("type"):
+                if msg_json.get("type") == "new_opportunity":
+                    if opportunity_callback is not None:
+                        opportunity = Opportunity.model_validate(
+                            msg_json.get("opportunity")
+                        )
+                        if opportunity:
```

Comment:
I don't think opportunity can be None here. model_validate throws an exception if invalid